### ISSUE TITLE
style disappearing on hard refresh / new load with nextJS

### ISSUE THREAD

**Issue by joelrozen** [chakra-ui/chakra-ui#2226]
# Bug report

## Describe the bug

Really weird, as I can't reproduce this on any other component or in any other project, I'm just hoping someone has seen something like this before and can point me in the right direction.

## To reproduce

I have:

```          
<Box height='250px' width='320px'>
    <Image
          src={post.fields.heroImage?.fields?.file?.url}
          objectFit='cover'
          borderRadius='15px'
          width='100%'
          height='100%'
          alt={post.fields?.heroImage?.fields?.title}
          // TODO: Fallback SRC fallbackSrc='TODO'
    />
</Box>
```

When I make any changes to either the parent Box or the Image, and nextJS fast reloads, it works perfectly. As soon as I hard refresh the page, the "height" and "width" attributes on the <Box> have completely disappeared, and thus the Box ends up taking the original dimensions of the Image. 

Inspecting dev tools shows after making *any* change:
BOX: 
![image](https://user-images.githubusercontent.com/20792877/95802498-5680fe00-0d49-11eb-961d-7d3cab2ea6dd.png)

IMAGE:
![image](https://user-images.githubusercontent.com/20792877/95802518-639ded00-0d49-11eb-9545-4e3a44a93cb6.png)

As soon as I reload the page, the class name of the BOX goes to `css-0` instead of the random css class it generated on the fast reload.

## Minimal reproduction

I can't really replicate this anywhere else, or even in isolation :(

## Expected behavior

the css class should remain!

## System information

- OS: [e.g. macOS, Windows] All
- Browser (if applies): [e.g. Chrome, Safari] All
- Version of @chakra-ui/core: [e.g. 1.0.0-rc.3] 0.8.0
- Version of Node.js: [e.g. 12.11.1] 12.18.3

## Additional context

My entire component looks like this - assuming it might have some weird bug in my parent(s)? 
Also I tried removing chakra and canning node_modules and re installing but no luck

```
import { Box, Heading, Image, PseudoBox, Tag, Text } from '@chakra-ui/core'
import { PostDisplayTypes } from '../types/postTypes'
import makeDate from '../utils/makeDate'
import shorten from '../utils/shortenSentence'
import { Link } from './Link'

const Card = ({ post }) => {
  return (
    <PseudoBox
      as='article'
      position='relative'
      height={['594px', '575px']}
      width={['321px', '350px']}
      borderRadius='15px'
      px='15px'
      pt='15px'
      m='20px'
      boxShadow='0px 5px 10px rgba(0, 0, 0, 0.16)'
      _hover={{
        boxShadow: '2px 10px 20px rgba(0, 0, 0, 0.32)',
      }}
      transition='all 0.2s ease'
    >
      <Link
        href={`/blog/${post.fields.slug}`}
        _hover={{ textDecoration: 'none' }}
        passHref
      >
        <Box>
          <Text
            fontSize='sm'
            color='cyan.700'
            pb='15px'
            fontWeight='bold'
            display='flex'
            justifyContent='space-between'
          >
            <Box>{makeDate(post.sys.createdAt)} </Box>
            <Box color='purple.700'>
              {PostDisplayTypes[post.sys.contentType.sys.id]}
            </Box>
          </Text>
          <Heading
            as='h3'
            fontSize='2xl'
            textAlign='left'
            minHeight='4em'
            maxHeight='4em'
          >
            {shorten(post.fields?.title, 75)}
          </Heading>
          <Text pb='15px' minHeight='9em' maxHeight='9em'>
            {shorten(post.fields.synopsis, 195)}
          </Text>
          <Box height='250px' width='320px'>
            <Image
              src={post.fields.heroImage?.fields?.file?.url}
              objectFit='cover'
              borderRadius='15px'
              width='100%'
              height='100%'
              alt={post.fields?.heroImage?.fields?.title}
              // TODO: Fallback SRC fallbackSrc='TODO'
            />
          </Box>
        </Box>
      </Link>

      <Box bottom='15px' m='15px 15px' position='absolute'>
        {post.fields?.tags?.slice(0, 2).map((tag) => (
          <Link
            href={`/tags/${tag.fields?.slug}`}
            _hover={{ textDecoration: 'none' }}
            passHref
          >
            <Tag
              variantColor='purple'
              rounded='full'
              variant='solid'
              mr='15px'
              boxShadow='2px 4px 10px rgba(0, 0, 0, 0.5)'
              _hover={{
                boxShadow: '4px 8px 20px rgba(0, 0, 0, 0.5)',
                bottom: '18px',
              }}
              transition='all 0.2s ease'
            >
              {tag.fields.name}
            </Tag>
          </Link>
        ))}
      </Box>
    </PseudoBox>
  )
}

export default Card
```

**Comment by segunadebayo** [comment]
I think this is related to the `Image` component.

If you're using version `1.0.0-rc.X`, kindly use the `Img` component instead.

**Comment by segunadebayo** [comment]
Also, I can't seem to reproduce as well. When there's a repro, I'll re-open it.

Thanks

**Comment by Jackbaude** [comment]
Im having this same issue, but only on a build version of a nextjs server.
Here is what it looks like on the first load:
![image](https://user-images.githubusercontent.com/25142905/110162290-27306380-7db4-11eb-99a5-a1516263b948.png)
And then after refresh
![image](https://user-images.githubusercontent.com/25142905/110162309-2c8dae00-7db4-11eb-8dc6-ec71d58aa352.png)

It also breaks when I change color mode from light
![image](https://user-images.githubusercontent.com/25142905/110162580-8bebbe00-7db4-11eb-8b99-c8d24e571c4f.png)

...to dark
![image](https://user-images.githubusercontent.com/25142905/110162659-a1f97e80-7db4-11eb-8596-a6da89907a5d.png)

Here is what my `_document.tsx` looks like, the only stuff that would mattter at least
```tsx
                <ColorModeScript/>
                <Main/>
                <NextScript/>
                </body>
            </Html>
```

And here is my `_app.tsx`
```tsx
            <ChakraProvider resetCSS theme={theme}>
                <Layout>
                    <Component {...pageProps} />
                </Layout>
            </ChakraProvider>
```
Any help with this would be appreciated, ive already asked the discord but no one seems to know either.

**Comment by binajmen** [comment]
Hi @Jackbaude @joelrozen @segunadebayo 

Did you find a solution to that issue ? I have exactly the same issue with NextJS + Chakra :(

![IMG_5170](https://user-images.githubusercontent.com/15611419/114090090-43de2080-98b7-11eb-8038-cc75a526099d.PNG)

I have this Drawer working in dev, but when running the build/prod version, I miss the style:

![IMG_5168](https://user-images.githubusercontent.com/15611419/114090178-5fe1c200-98b7-11eb-983f-e3c6e4e44923.PNG)

Hitting hard refresh make it works.

I noticed the following:

![Screenshot 2021-04-08 at 22 12 36](https://user-images.githubusercontent.com/15611419/114090399-a9321180-98b7-11eb-89fb-f53dd640d03c.png)

![Screenshot 2021-04-08 at 22 12 56](https://user-images.githubusercontent.com/15611419/114090408-ac2d0200-98b7-11eb-84e5-8f7f4aac874f.png)

As you yan see in the right side panel, it's like the styles classes css-XXXXX are no more present. Unfortunately, I cannot share the project repo as it is not mine, and I wasn't able to reproduce it outside of my current project.

**Comment by Jackbaude** [comment]
@binajmen 
Yes I did. My issue was due to the slate js editor. What I had to do was dynamically load that component wirh next js and then it worked. This is likely not the solution to your problem though.

**Comment by joelrozen** [comment]
From memory my issue was stale builds and browser caching. I deleted my dist folder and ran dev again and it was fine after that.

**Comment by binajmen** [comment]
In my case, this is happening in prod but not in dev (strangely enough..)

**Comment by ThomasClaes-r0664186** [comment]
Hi @binajmen,

Have you found a solution, i seem to be having a similar problem. 

As you can see in the video below, the style works for a while, but then suddenly it dissapears on some pages. 

However after a refresh the style seems to be working just fine again?
Any ideas on this? :)

https://user-images.githubusercontent.com/28530291/115574314-cd332100-a2c1-11eb-87cf-1ca4525d7657.mp4

**Comment by binajmen** [comment]
Hi @ThomasClaes-r0664186,

It was really specific to my setup: it was a combo of (1) a third-party library that was (badly) using `emotion`(Chakra UI is based on emotion) and (2) NextJS that performs prefetch of links. The two combined, the prefetching was causing `emotion` to reload, causing an unsync of the styles. My solution was to throw the third-party library (react google maps autocomplete and build my own solution).

I doubt this is the same situation, but perhaps it could help you to narrow down the problem.

**Comment by ThomasClaes-r0664186** [comment]
> Hi @ThomasClaes-r0664186,
> 
> It was really specific to my setup: it was a combo of (1) a third-party library that was (badly) using `emotion`(Chakra UI is based on emotion) and (2) NextJS that performs prefetch of links. The two combined, the prefetching was causing `emotion` to reload, causing an unsync of the styles. My solution was to throw the third-party library (react google maps autocomplete and build my own solution).
> 
> I doubt this is the same situation, but perhaps it could help you to narrow down the problem.

Thanks for the response, i've also used the react-google-maps so maybe i'll try looking there. 😁

**Comment by MalikBagwala** [comment]
My solution was to load the problematic component **without SSR**

```jsx
import dynamic from 'next/dynamic'

const DynamicComponent = dynamic(() => import('../components/hello'))

function Home() {
  return (
    <div>
      <Header />
      <DynamicComponent />
      <p>HOME PAGE is here!</p>
    </div>
  )
}

export default Home
```

**Comment by harrysoer** [comment]
I also had this problem with my component that is conditionally being rendered. The FIX I did for here is the same with @Jackbaude by using NextJS's `dynamic` import and viola, my styling is not broken anymore. But TBH I'm not really sure what's causing this broken styling.... I'm guessing that this was caused by emotion's build for SSR 🤔 If anyone found out, I'm interested to have a discussion about it 👍

**Comment by Tarunkr89** [comment]
this package works for me.
@zeit/next-css

// next.config.js
const withCSS = require('@zeit/next-css')
module.exports = withCSS()

**Comment by jeroenvanhattem** [comment]
Same issue. The solution provided by @Jackbaude and @harrysoer fixed the issue for me in 1 place. Weird thing is that it doens't work with a different component in the same project. If someone finds out what's causing it, I'd love to know

**Comment by NitinSirsath** [comment]
if you are using styled-components then do the following and you are good to go
Step 1: npm install --save-dev babel-plugin-styled-components
Step 2: Create a file [.babelrc] in the root directory of next project
Step 3: Paste Following Code to .babelrc File:-

             {
                 "presets": ["next/babel"],
                 "plugins": [["styled-components", { "ssr": true }]]
             }

Step 4: Restart the next server

**Comment by GavintheAmateur** [comment]
I met the same issue... I guess it's some conflict on the frameworks I used: Next and MUI, then Chakra-ui added to play around.  I could see the class name became "css-0", which indicates the styles on Chakra were lost. I hope someone walks me through a bit here how the style compilation works under the hood and how the conflict was exactly caused.

**Comment by ryanSN** [comment]
I was having a similar issue.
I had a `Flex` that contained two `Box` components one had content and the other contained a `next/image` that was responsive. Removing the image seemed to fix the page, but it wasn't the main issue. 

Navigating to the page would keep the styles, but a refresh  / hot reload would cause one box that contained a background on the page to lose all its styles.
Removing the flex and its contents _Fixed_ the issue.
I noticed I had two un-needed containing `Box` elements wrapping my content and the flex. Removal of those seemed to address the issue, then the flex styles seemed to bleed onto `Box`'s that didn't have any reason to be in a flex. There was another Box that wrapped the flex and the body contents below (this was unneeded and foolish on my part).

if anyone sees the hard refresh causes the `css-0` and you do not have any extra emotion imports, I would look at your component structure. 

Version saw the issue Chakra-ui 1.8.6

tl;dr; 
Flex / Grid where causing styles to be lost, but in the end, it was too many wrapping `Box` elements.

**Comment by Weave-Trix** [comment]
> 

this works for me... thanks a lot!

**Comment by Weave-Trix** [comment]
> this package works for me. @zeit/next-css
> 
> // next.config.js const withCSS = require('@zeit/next-css') module.exports = withCSS()

This works for me..Thanks a lot!!!

**Comment by Weave-Trix** [comment]
> if you are using styled-components then do the following and you are good to go Step 1: npm install --save-dev babel-plugin-styled-components Step 2: Create a file [.babelrc] in the root directory of next project Step 3: Paste Following Code to .babelrc File:-
> 
> ```
>          {
>              "presets": ["next/babel"],
>              "plugins": [["styled-components", { "ssr": true }]]
>          }
> ```
> 
> Step 4: Restart the next server

This works for me. Thanks a lot @NitinSirsath !!!

**Comment by Pedr0Leite** [comment]
> if you are using styled-components then do the following and you are good to go Step 1: npm install --save-dev babel-plugin-styled-components Step 2: Create a file [.babelrc] in the root directory of next project Step 3: Paste Following Code to .babelrc File:-
> 
> ```
>          {
>              "presets": ["next/babel"],
>              "plugins": [["styled-components", { "ssr": true }]]
>          }
> ```
> 
> Step 4: Restart the next server

This worked for me!! Thank you so much!

**Comment by sandersgutierrez** [comment]
@NitinSirsath This worked perfectly for me! Thank you so much!

> if you are using styled-components then do the following and you are good to go Step 1: npm install --save-dev babel-plugin-styled-components Step 2: Create a file [.babelrc] in the root directory of next project Step 3: Paste Following Code to .babelrc File:-
> 
> ```
>          {
>              "presets": ["next/babel"],
>              "plugins": [["styled-components", { "ssr": true }]]
>          }
> ```
> 
> Step 4: Restart the next server

### Share my standard output

```shell
> next dev

ready - started server on 0.0.0.0:3000, url: http://localhost:3000
info  - Disabled SWC as replacement for Babel because of custom Babel configuration "babel.config.json" https://nextjs.org/docs/messages/swc-disabled
...
...
event - compiled client and server successfully in 1888 ms (262 modules)
```

**Comment by Anboias** [comment]
https://github.com/chakra-ui/chakra-ui/issues/2226#issuecomment-1226906513

The `.babelrc` config might not be needed anymore, adding `styledComponents: true` to `next.config.js` did the job for me.

```
const nextConfig = {
  ...
  compiler: {
    ...
    styledComponents: true,
  },
};
```

**Comment by danstacyalltech** [comment]
> [#2226 (comment)](https://github.com/chakra-ui/chakra-ui/issues/2226#issuecomment-1226906513)
> 
> The `.babelrc` config might not be needed anymore, adding `styledComponents: true` to `next.config.js` did the job for me.
> 
> ```
> const nextConfig = {
>   ...
>   compiler: {
>     ...
>     styledComponents: true,
>   },
> };
> ```

This worked for me. Thank you.
